notification/sse.go

```diff
@@ -0,0 +1,104 @@
+package notification
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+	"strings"
+
+	"github.com/linksmart/thing-directory/common"
+)
+
+const (
+	QueryParamType    = "type"
+	QueryParamFull    = "full"
+	HeaderLastEventID = "Last-Event-ID"
+)
+
+type SSEAPI struct {
+	controller  NotificationController
+	contentType string
+}
+
+func NewHTTPAPI(controller NotificationController, version string) *SSEAPI {
+	contentType := "text/event-stream"
+	if version != "" {
+		contentType += ";version=" + version
+	}
+	return &SSEAPI{
+		controller:  controller,
+		contentType: contentType,
+	}
+
+}
+
+func (a *SSEAPI) SubscribeEvent(w http.ResponseWriter, req *http.Request) {
+	eventTypes, full, err := parseQueryParameters(req)
+	if err != nil {
+		common.ErrorResponse(w, http.StatusBadRequest, err)
+	}
+	flusher, ok := w.(http.Flusher)
+	if !ok {
+		common.ErrorResponse(w, http.StatusInternalServerError, "Streaming unsupported")
+		return
+	}
+	w.Header().Set("Content-Type", a.contentType)
+
+	messageChan := make(chan Event)
+
+	lastEventID := req.Header.Get(HeaderLastEventID)
+	a.controller.subscribe(messageChan, eventTypes, full, lastEventID)
+
+	go func() {
+		<-req.Context().Done()
+		a.controller.unsubscribe(messageChan)
```

Comment:
Does this close `messageChan` so that the following loop exits?

Reply:
good catch. Now it is closed in the handler function of controller